src/Network/Transport/TCP.hs

```diff
@@ -517,20 +519,25 @@ data TransportInternals = TransportInternals
 --------------------------------------------------------------------------------
 
 -- | Create a TCP transport
-createTransport :: N.HostName
-                -> N.ServiceName
+createTransport :: N.HostName    -- ^ Bind host name.
+                -> N.ServiceName -- ^ Bind port.
+                -> (N.ServiceName -> (N.HostName, N.ServiceName))
+                   -- ^ External address host name and port, computed from the
+                   --   actual bind port (must be lazy in it).
```

Comment:
We could get rid of this laziness requirement if we make the `transportHost` and `transportPort` fields of `TCPTransport` non-strict. `createTransportExposeInternals` could compute these things after the port is determined (after `forkServer`) but pass them lazily to the `TCPTransport` constructor.